### PR TITLE
feat(soql): use sobject cache

### PR DESCRIPTION
**Pull request by jgellin-sf**
### What does this PR do?
This PR changes the SOQL extension so that the SOQL Editor caches SObjects that have been retrieved from the connected org. It also introduces a new message, `connection_changed`, which allows the editor to signal to the UI that the connection has changed so that it can react appropriately.

### What issues does this PR fix or reference?
@W-8253021@ @W-8207695@

### Functionality Before
SOQL Editor makes web service request each time the list of SObjects or metadata for a particular SObject is needed. When connection changes, the UI doesn't know and can't refresh its SObject metadata.

### Functionality After
SOQL Editor makes web service request only when cache cannot return the requested information. When connection changes, the UI is sent a message and can react appropriately.

### CODE REVIEW

File: packages/salesforcedx-vscode-soql/src/editor/soqlEditorInstance.ts [hunk 1]
```diff
@@ -52,6 +53,43 @@ async function withSFConnection(f: (conn: Connection) => void): Promise<void> {
   }
 }
 
+class ConnectionChangedListener {
```
Comment (lcampos) [inline packages/salesforcedx-vscode-soql/src/editor/soqlEditorInstance.ts:56]: Is this PR only taking care of the connection handling ? I don't see the editor handling the event/message sent when switching connections (`onConnectionChange`).

Comment (jgellin-sf) [reply packages/salesforcedx-vscode-soql/src/editor/soqlEditorInstance.ts:56]: Yes. The editor code is here: https://github.com/forcedotcom/soql-tooling/pull/60

Comment (lcampos) [reply packages/salesforcedx-vscode-soql/src/editor/soqlEditorInstance.ts:56]: Cool 👍

**Review by lcampos — APPROVED** [review]

File: packages/salesforcedx-vscode-soql/package.json [hunk 2]
```diff
@@ -21,10 +21,10 @@
...
   "dependencies": {
     "@salesforce/core": "^2.12.1",
-    "@salesforce/soql-builder-ui": "0.0.10",
-    "@salesforce/soql-data-view": "0.0.5",
-    "@salesforce/soql-language-server": "0.2.1",
-    "@salesforce/soql-tmlanguage": "0.1.1",
+    "@salesforce/soql-builder-ui": "^0.0.20",
+    "@salesforce/soql-data-view": "^0.0.5",
+    "@salesforce/soql-language-server": "^0.2.5",
+    "@salesforce/soql-tmlanguage": "^0.1.1",
```
Comment (dehru) [inline packages/salesforcedx-vscode-soql/package.json:27]: @jgellin-sf I've got these dependencies locked down.  And updated to the latest.  Can you match and remove the ^?

https://github.com/forcedotcom/salesforcedx-vscode/pull/2695/files

Comment (jgellin-sf) [reply packages/salesforcedx-vscode-soql/package.json]: @dehru As you wish. See `chore(soql): lock down dependency versions`.

Comment (jgellin-sf) [reply packages/salesforcedx-vscode-soql/package.json]: @dehru CircleCI is complaining about `@salesforce/soql-language-server 0.2.7`. Is this version not published yet?

**Review by jonnyhork — APPROVED** [review]
QA Passed

**Pull request merged**